src/attr.rs

```diff
@@ -0,0 +1,205 @@
+/// assert_eq!(format!("{}", attr), "invalid=id=\"foo\"");
+/// ```
+///
+/// ```rust
+/// let attr = hamlet::Attribute::new("checked", "");
+/// assert_eq!(format!("{}", attr), "checked");
+/// ```
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Attribute<'a> {
+    pub name: Cow<'a, str>,
+    pub value: Cow<'a, str>,
+}
+
+impl<'a> Attribute<'a> {
+    /// Create an attribute, useful to avoid having to convert strings to
+    /// `Cow<str>` yourself.
+    ///
+    /// Generally this shouldn't be used directly by end users, it's likely
+    /// that there are builder APIs or macros available that make attribute
+    /// construction easier, for example the modification methods on
+    /// [`AttributeSet`](struct.AttributeSet.html#methods) or the
+    /// [`attr_set!`](macro.attr_set!.html) macro.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use std::borrow::Cow;
+    /// let foo = "foo".to_owned();
+    /// let foo2 = foo.clone();
+    /// assert_eq!(
+    ///     hamlet::Attribute::new("id", foo),
+    ///     hamlet::Attribute {
+    ///         name: Cow::Borrowed("id"),
+    ///         value: Cow::Owned(foo2),
+    ///     });
+    /// ```
+    pub fn new<N, V>(name: N, value: V) -> Attribute<'a>
+        where N: Into<Cow<'a, str>>,
+              V: Into<Cow<'a, str>>
+    {
+        Attribute {
+            name: name.into(),
+            value: value.into(),
+        }
+    }
+}
+
+impl<'a> fmt::Display for Attribute<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if self.value == "" {
+            write!(f, "{}", self.name)
+        } else {
+            write!(f, "{}=\"{}\"", self.name.as_ref(), Escaped(&self.value))
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+/// A set of [`Attribute`s](./struct.Attribute.html).
+///
+/// This is stored as a plain slice instead of a set as in most cases it will
+/// be a small collection over which linear search will be more efficient.
+///
+/// Generally end users shouldn't construct this struct directly, it's expected
+/// that there will be builder APIs or macros available to make construction
+/// easier, for example the provided [`attr_set!`](macro.attr_set!.html) macro.
+pub struct AttributeSet<'a>(pub Cow<'a, [Attribute<'a>]>);
+
+impl<'a> AttributeSet<'a> {
+    /// Try and get a reference to an existing attribute.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # #[macro_use] extern crate hamlet;
+    /// # fn main() {
+    /// let attrs = attr_set!(id = "foo");
+    /// assert_eq!(
+    ///     attrs.get_attr("id"),
+    ///     Some(&hamlet::Attribute::new("id", "foo")));
+    /// # }
+    /// ```
+    ///
+    /// ```rust
+    /// # #[macro_use] extern crate hamlet;
+    /// # fn main() {
+    /// let attrs = attr_set!(id = "foo");
+    /// assert_eq!(
+    ///     attrs.get_attr("class"),
+    ///     None);
+    /// # }
+    /// ```
+    pub fn get_attr<S>(&self, name: S) -> Option<&Attribute<'a>>
```

Comment:
What's your opinion about changing this to `get_attr_val` for directly fetching the value?